### PR TITLE
Overhaul token cursors

### PR DESCRIPTION
**Pull request by nnethercote**
Some nice cleanups here.

r? @davidtwco

### CODE REVIEW

**Comment by nnethercote** [comment]
I saw small but consistent icount improvements locally:
Screenshot from 2024-12-12 07-21-37

The most likely explanation when local wins aren't seen on CI is that PGO is involved. But not every platform has PGO applied, and the improvements here are also to readability, so I think this is still worth going forward.

**Comment by nnethercote** [comment]
@petrochenkov: I know your time is limited at the moment. Unlike PRs such as #133436 this PR is not that complicated, so I will ask someone else to review.

r? @davidtwco

File: src/tools/rustfmt/src/macros.rs [hunk 1]
```diff
@@ -1237,10 +1237,10 @@ impl<'a> MacroParser<'a> {
                 span,
             },
             _,
-        )) = self.toks.look_ahead(0)
+        )) = self.iter.peek()
```
Comment (ytmimi) [inline src/tools/rustfmt/src/macros.rs:1240]: I see that we're replacing calls to `.look_ahead()` with `.peek()`. Is `.peek()` equivalent to both `.look_ahead(0)` here and `.look_ahead(1)` above in `MacroParser::parse`?

Comment (nnethercote) [reply src/tools/rustfmt/src/macros.rs:1240]: Yes. Full details are in the individual commit messages.

Comment (ytmimi) [reply src/tools/rustfmt/src/macros.rs:1240]: Interesting that it doesn't matter if you use `0`, `1`, or `2` as the look ahead, and that things only start failing if you choose `3` or more. I know you mentioned it's due to the structure of the code, but I'm wondering if you could elaborate a little more to help me fully understand.

Comment (nnethercote) [reply src/tools/rustfmt/src/macros.rs:1240]: AIUI, the existing code says "If there are at least two tokens remaining, try to parse a branch". The new code says "if there is at least one token remaining, try to parse a branch". `parse_branch` has three `self.toks.next()?` calls in it, and all three must succeed for the function to succeed, i.e. there must actually be at least three tokens remaining.

**Comment by spastorino** [comment]
@bors r-

There are some conflicts, please r=me after addressing them.

**Comment by nnethercote** [comment]
I rebased.

@bors r=spastorino rollup=maybe

**Pull request merged**